File: public/js/core3_example_ui/groups.js
/**
  * Copyright 2017 Interel
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  * http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
define(["utils"],
    function (utils) {
        function Groups() {}

        Groups.clearForm = function (form) {
            form.get(0).reset();
        };

        Groups.createHandler = function (form, e) {
            var nameControl = $("#c3eu-group-name");
            var shortNameControl = $("#c3eu-group-short-name");
            var itemsTypeControl = $("#c3eu-group-items-type");

            var name = nameControl.val();
            var shortName = shortNameControl.val();
            var itemsType = itemsTypeControl.find("option:selected").val();

            var errors = false;
            if (!name) {
                errors = true;
                nameControl.addClass("uk-form-danger");
            } else {
                nameControl.removeClass("uk-form-danger");
            }

            if (!shortName) {
                errors = true;
                shortNameControl.addClass("uk-form-danger");
            } else {
                shortNameControl.removeClass("uk-form-danger");
            }

            if (!itemsType) {
                errors = true;
                itemsTypeControl.addClass("uk-form-danger");
            } else {
                itemsTypeControl.removeClass("uk-form-danger");
            }


            if (!errors) {
                UIkit.notification("<div uk-spinner></div> Updating...", {status: "primary", pos: "top-right"});

                utils.post(
                    "/groups/create",
                    {
                        "name": name,
                        "shortName": shortName,
                        "itemsType": itemsType,
                        "items": []
                    }
                ).done(function (createResult) {
                    if (createResult.wasSuccessful) {
                        $("#c3eu-group-grid").prepend(createResult.data.html);

                        UIkit.notification.closeAll();
                        UIkit.notification("<span uk-icon='icon: check'></span> Done.", {
                            status: "success",
                            pos: "top-right"
                        });
                    } else {
                        console.error("Operation failed: [" + createResult.message + "]");
                        UIkit.notification("<span uk-icon='icon: warning'></span> Failed to create group.", {
                            status: "danger",
                            pos: "top-right"
                        });
                    }
                }).fail(function (xhr, status, error) {
                    console.error("Operation failed with status [" + xhr.status + "] and message [" + error + "]");
                    UIkit.notification("<span uk-icon='icon: warning'></span> Failed to create group [" + xhr.status + "].", {
                        status: "danger",
                        pos: "top-right"
                    });
                });
            } else {
                e.stopPropagation();
                e.preventDefault();
            }
        };

        Groups.deleteHandler = function (e) {
            var groupContainer = $(e.currentTarget).closest(".c3eu-group-container");
            var objectData = groupContainer.find(".c3eu-group-object-data");
            var groupName = groupContainer.find(".c3eu-group-name").text();

            var id = objectData.attr("data-id");
            var revision = objectData.attr("data-revision");
            var revisionNumber = objectData.attr("data-revisionNumber");

            UIkit.modal.confirm("Remove group [" + groupName + "]?").then(
                function () {
                    UIkit.notification("<div uk-spinner></div> Updating...", {status: "primary", pos: "top-right"});

                    utils.post(
                        "/groups/delete",
                        {
                            "groupID": id,
                            "revision": revision,
                            "revisionNumber": revisionNumber
                        }
                    ).done(function (result) {
                        if (result.wasSuccessful) {
                            groupContainer.parent().remove();

                            UIkit.notification.closeAll();
                            UIkit.notification("<span uk-icon='icon: check'></span> Done.", {status: "success", pos: "top-right"});
                        } else {
                            console.error("Operation failed: [" + result.message + "]");
                            UIkit.notification("<span uk-icon='icon: warning'></span> Failed to delete group.", {status: "danger", pos: "top-right"});
                        }
                    }).fail(function (xhr, status, error) {
                        console.error("Operation failed with status [" + xhr.status + "] and message [" + error + "]");
                        UIkit.notification("<span uk-icon='icon: warning'></span> Failed to delete group [" + xhr.status + "].", {
                            status: "danger",
                            pos: "top-right"
                        });
                    });
                },
                function () {
                });
        };

        Groups.editHandler = function (e) {
            var groupContainer = $(e.currentTarget).closest(".c3eu-group-container");

            var name = groupContainer.find(".c3eu-group-name");

            var originalName = name.text();

            var groupButtonContainer = groupContainer.find(".c3eu-group-button-container");
            var groupEditButton = groupButtonContainer.find(".c3eu-group-edit");
            var groupDeleteButton = groupButtonContainer.find(".c3eu-group-delete");

            var groupSaveButton = $("<a></a>", {
                title: "Save",
                "class": "uk-icon-button uk-text-success c3eu-group-save",
                "uk-icon": "icon: check",
                "uk-tooltip": "",
                click: function () {
                    groupButtonContainer.find(".c3eu-group-save").addClass("uk-disabled");

                    var objectData = groupContainer.find(".c3eu-group-object-data");

                    var id = objectData.attr("data-id");
                    var revision = objectData.attr("data-revision");
                    var revisionNumber = objectData.attr("data-revisionNumber");

                    UIkit.notification("<div uk-spinner></div> Updating...", {status: "primary", pos: "top-right"});

                    utils.post(
                        "/groups/update",
                        {
                            "groupID": id,
                            "revision": revision,
                            "revisionNumber": revisionNumber,
                            "name": name.text()
                        }
                    ).done(function (updateResult) {
                        if (updateResult.wasSuccessful) {
                            groupContainer.parent().replaceWith(updateResult.data.html);

                            UIkit.notification.closeAll();
                            UIkit.notification("<span uk-icon='icon: check'></span> Done.", {
                                status: "success",
                                pos: "top-right"
                            });
                        } else {
                            console.error("Operation failed: [" + updateResult.message + "]");
                            UIkit.notification("<span uk-icon='icon: warning'></span> Failed to update group.", {
                                status: "danger",
                                pos: "top-right"
                            });
                        }
                    }).fail(function (xhr, status, error) {
                        console.error("Operation failed with status [" + xhr.status + "] and message [" + error + "]");
                        UIkit.notification("<span uk-icon='icon: warning'></span> Failed to update group [" + xhr.status + "].", {
                            status: "danger",
                            pos: "top-right"
                        });
                    });
                }
            });

            var groupCancelButton = $("<a></a>", {
                title: "Cancel",
                "class": "uk-icon-button uk-text-danger c3eu-group-cancel",
                "uk-icon": "icon: close",
                "uk-tooltip": "",
                click: function () {
                    name.text(originalName);

                    name.removeClass("c3eu-editable");
                    name.attr("contentEditable", false);
                    groupEditButton.removeClass("uk-hidden");
                    groupDeleteButton.removeClass("uk-hidden");
                    groupContainer.find(".c3eu-group-save").remove();
                    groupContainer.find(".c3eu-group-cancel").remove();
                }
            });

            name.addClass("c3eu-editable");
            name.attr("contentEditable", true);
            groupEditButton.addClass("uk-hidden");
            groupDeleteButton.addClass("uk-hidden");
            groupButtonContainer.append(groupSaveButton);
            groupButtonContainer.append(groupCancelButton);
        };

        Groups.prototype.init = function () {
            var grid = $("#c3eu-group-grid");

            $("body").on("hide", "[uk-modal]", function (e) {
                Groups.clearForm($(e.currentTarget).find("form"))
            });

            $(".c3eu-group-add-create").click(function (e) {
                Groups.createHandler($("#c3eu-group-add-form"), e);
            });

            grid.on("click", ".c3eu-group-delete", Groups.deleteHandler);
            grid.on("click", ".c3eu-group-edit", Groups.editHandler);
        };

        return new Groups();
    }
);

File: app/controllers/Groups.scala
/**
  * Copyright 2017 Interel
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  * http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
package controllers

import javax.inject.Inject

import containers.UserData
import core3.config.StaticConfig
import core3.database.ObjectID
import core3.database.containers.core
import core3.database.dals.DatabaseAbstractionLayer
import core3.http.controllers.local.ClientController
import core3.http.requests.WorkflowEngineConnection
import core3.http.responses.GenericResult
import core3.workflows.{NoWorkflowParameters, WorkflowRequest}
import core3.workflows.definitions._
import play.api.Environment
import play.api.cache.SyncCacheApi
import play.api.data.Forms._
import play.api.data._
import play.api.libs.json.Json
import play.filters.csrf.CSRF

import scala.concurrent.{ExecutionContext, Future}

class Groups @Inject()(engineConnection: WorkflowEngineConnection, cache: SyncCacheApi, db: DatabaseAbstractionLayer)
  (implicit ec: ExecutionContext, environment: Environment)
  extends ClientController(cache, StaticConfig.get.getConfig("security.authentication.clients.LocalUIExample"), db) {

  //Groups page handler - retrieves all groups, sorts them by creation date and renders them
  def page() = AuthorizedAction(
    "c3eu:view",
    okHandler = { (request, user) =>
      implicit val r = request
      implicit val token = CSRF.getToken

      val userData = UserData(user)
      for {
        result <- engineConnection.post(
          user,
          WorkflowRequest(
            SystemQueryGroups.name,
            NoWorkflowParameters().asJson
          )
        )
      } yield {
        if (result.wasSuccessful) {
          val groups = result.data.map {
            output =>
              (output \ "groups").as[Vector[core.Group]]
          }.getOrElse(Seq.empty).sortWith(
            (a, b) =>
              a.created.isAfter(b.created)
          )

          val userData = UserData(user)

          Ok(views.html.groups.page("Groups", userData, groups))
        } else {
          Ok(views.html.system.error("Groups", result.message.getOrElse("No message provided"), Some(userData)))
        }
      }
    }
  )

  //Groups creation handler - processes the supplied data, sends a request to the workflow engine and renders the new object, ready to be inserted into the DOM
  def create() = AuthorizedAction(
    "c3eu:edit",
    okHandler = { (request, user) => implicit val r = request
      Groups.Forms.create.bindFromRequest.fold(
        form =>
          Future.successful(Ok(GenericResult(wasSuccessful = false, message = Some(s"Failed to validate input: [${form.errors}]")).asJson))
        ,
        params =>
          for {
            result <- engineConnection.post(
              user,
              WorkflowRequest(
                SystemCreateGroup.name,
                params.asJson,
                returnOutputData = true
              )
            )
          } yield {
            if (result.wasSuccessful) {
              val group = result.data match {
                case Some(data) => (data \ "add")(0).as[core.Group]
                case None => throw new RuntimeException(s"Invalid response received from service.")
              }

              val userData = UserData(user)

              Ok(
                GenericResult(
                  wasSuccessful = true,
                  message = None,
                  data = Some(
                    Json.obj(
                      "html" -> views.html.groups._group(group, userData, isNew = true).body
                    )
                  )
                ).asJson
              )
            } else {
              Ok(result.asJson)
            }
          }
      )
    }
  )

  //Groups deletion handler - processes the supplied data, sends a request to the workflow engine and responds with a generic result
  def delete() = AuthorizedAction(
    "c3eu:delete",
    okHandler = { (request, user) => implicit val r = request
      Groups.Forms.delete.bindFromRequest.fold(
        form =>
          Future.successful(Ok(GenericResult(wasSuccessful = false, message = Some(s"Failed to validate input: [${form.errors}]")).asJson))
        ,
        params =>
          for {
            result <- engineConnection.post(user, WorkflowRequest(SystemDeleteGroup.name, params.asJson))
          } yield {
            if (result.wasSuccessful) {
              Ok(GenericResult(wasSuccessful = true).asJson)
            } else {
              Ok(result.asJson)
            }
          }
      )
    }
  )

  //Groups update handler - processes the supplied data, sends a request to the workflow engine and renders the updated object, ready to be re-inserted into the DOM
  def update() = AuthorizedAction(
    "c3eu:edit",
    okHandler = { (request, user) => implicit val r = request
      Groups.Forms.update.bindFromRequest.fold(
        form =>
          Future.successful(Ok(GenericResult(wasSuccessful = false, message = Some(s"Failed to validate input: [${form.errors}]")).asJson))
        ,
        params =>
          for {
            result <- engineConnection.post(
              user,
              WorkflowRequest(
                SystemUpdateGroup.name,
                params.asJson,
                returnOutputData = true
              )
            )
          } yield {
            if (result.wasSuccessful) {
              val group = result.data match {
                case Some(data) => (data \ "update")(0).as[core.Group]
                case None => throw new RuntimeException(s"Invalid response received from service.")
              }

              val userData = UserData(user)

              Ok(
                GenericResult(
                  wasSuccessful = true,
                  message = None,
                  data = Some(
                    Json.obj(
                      "html" -> views.html.groups._group(group, userData, isNew = true).body
                    )
                  )
                ).asJson
              )
            } else {
              Ok(result.asJson)
            }
          }
      )
    }
  )
}

object Groups {
  object Forms {
    val create = Form(
      mapping(
        "shortName" -> nonEmptyText,
        "name" -> nonEmptyText,
        "items" -> list(uuid).transform[Vector[ObjectID]](_.toVector, _.toList),
        "itemsType" -> nonEmptyText
      )(SystemCreateGroup.SystemAddGroupParameters.apply)(SystemCreateGroup.SystemAddGroupParameters.unapply)
    )

    val delete = Form(
      mapping(
        "groupID" -> uuid,
        "revision" -> uuid,
        "revisionNumber" -> number
      )(SystemDeleteGroup.SystemDeleteGroupParameters.apply)(SystemDeleteGroup.SystemDeleteGroupParameters.unapply)
    )

    val update = Form(
      mapping(
        "groupID" -> uuid,
        "revision" -> uuid,
        "revisionNumber" -> number,
        "name" -> optional(nonEmptyText),
        "items" -> optional(list(uuid).transform[Vector[ObjectID]](_.toVector, _.toList))
      )(SystemUpdateGroup.SystemUpdateGroupParameters.apply)(SystemUpdateGroup.SystemUpdateGroupParameters.unapply)
    )
  }
}
